### ISSUE TITLE
Trying to add redir to the domain directive and slash at the end of the proxied host

### ISSUE THREAD

**Issue by anLizard** [lucaslorentz/caddy-docker-proxy#134]
In my use case, my proxied app seems to only work when going to `/syncthing/` or proxying `http://syncthing:8384/`.
Just `/syncthing` or `http://syncthing:8384` doesnt work. It needs the extra slash at the end.
I'm trying to add a redir and attach the slash, so I'd end up with this:

```
sub.example.com
	redir /syncthing /syncthing/ 302
	proxy /syncthing http://syncthing:8384/ {
		transparent
		without /syncthing
}
```
Sadly, I'm stuck and dont know how to continue with my docker-compose.yml.

```
  caddy:
    image: lucaslorentz/caddy-docker-proxy:0.3.4
    container_name: caddy-0
    command: -email mail@example.com -agree=true
    networks:
      - public
      - internal
    ports:
      - "80:80"
      - "443:443"
    restart: always
    volumes:
      - ./caddy:/root/.caddy
      - ./www:/var/www
      - /etc/localtime:/etc/localtime:ro
      - /var/run/docker.sock:/var/run/docker.sock
    env_file:
      - .env
    environment:
      # prehend raw Caddyfile content
      - CADDY_DOCKER_CADDYFILE_PATH=/root/.caddy/Caddyfile
    labels:
      # nonwww redirect
      - caddy_0=$MY_DOMAIN
      - caddy_0.redir=https://www.$MY_DOMAIN{uri}
      # serve static content
      - caddy_1=www.$MY_DOMAIN
      - caddy_1.root=/var/www/$MY_DOMAIN/public

  syncthing:
    image: linuxserver/syncthing
    container_name: syncthing-0
    networks:
      - internal
      - public
    ports:
      #webUI
       #- 24100:8384
      #listening
      - 22000:22000
      #port discovery
      - 21027:21027/udp
    restart: always
    volumes:
      - ./syncthing/config:/config
      - ./syncthing/data:/data
      - /etc/localtime:/etc/localtime:ro
    environment:
      - UMASK_SET=<022>
      - PUID=1002
      - PGID=999
    labels:
      - caddy.redir=/syncthing /syncthing/ 302
      - caddy.address=shh.$MY_DOMAIN/syncthing
      - caddy.targetport=8384/

```

**Comment by lucaslorentz** [comment]
Currently, there is no support for configuring this part of the proxy: proxy **/syncthing** http://syncthing:8384/

But we have an option to generate the proxy by yourself. This should give you the caddyfile you want:
labels:
  - caddy=sub.example.com
  - caddy.redir=/syncthing /syncthing/ 302
  - caddy.proxy=/syncthing http://syncthing-0:8384/
  - caddy.transparent=
  - caddy.without=/syncthing

**Comment by anLizard** [comment]
I'm getting `421 Site sub.example.com is not served on this interface` on both `https://sub.example.com/syncthing` and `https://sub.example.com/syncthing/` now.

Is there any way I can check the in-memory Dockerfile?

**Comment by lucaslorentz** [comment]
Since version v0.3.5 (just released), you can do that configuration with:
- caddy.redir=/syncthing /syncthing/ 302
- caddy.address=sub.example.com
- caddy.sourcepath=/syncthing
- caddy.targetport=8384
- caddy.transparent=
- caddy.without=/syncthing

**Comment by lucaslorentz** [comment]
Yes, in memory caddyfile is logged in caddy container. Check the logs there.

**Comment by anLizard** [comment]
Many thanks. Using your [most recently provided labels](https://github.com/lucaslorentz/caddy-docker-proxy/issues/134#issuecomment-589236326), I get this in the dockerfile, but still experience 421 on both uris

```
shh.lzrd.org {
  proxy / 172.20.0.5:8384 172.21.0.3:8384
  redir /syncthing /syncthing/ 302
  sourcepath /syncthing
  transparent
  without /syncthing
}
```

I believe a simple redirect from /syncthing to /syncthing/ could also suffice for my needs

**Comment by lucaslorentz** [comment]
Ok, I think that's because depending on your domain, caddy by default serve it on port 2015.
Check: https://caddyserver.com/v1/docs/http-caddyfile

You can try to change your address to:
```
caddy.address=http://sub.example.com https://sub.example.com
```

**Comment by lucaslorentz** [comment]
> Many thanks. Using your [most recently provided labels](https://github.com/lucaslorentz/caddy-docker-proxy/issues/134#issuecomment-589236326), I get this in the dockerfile, but still experience 421 on both uris
> 
> ```
> sub.example.com {
>   proxy / 172.20.0.5:8384 172.21.0.3:8384
>   redir /syncthing /syncthing/ 302
>   sourcepath /syncthing
>   transparent
>   without /syncthing
> }
> ```
> 
> I believe a simple redirect from /syncthing to /syncthing/ could also suffice for my needs

Your caddyfile should be a bit different for those labels. Update your image to `lucaslorentz/caddy-docker-proxy:0.3.5`. sourcepath was introduced on that version

**Comment by anLizard** [comment]
I'm using my own domain, acme http-01 challenges and static content were working as expected on 80 and 443.
I temporarily mapped 2015 to a different port for testing, but all requests there resulted in 404.

After updating to 0.3.5, only port 2015 is working and firefox fails to connect to via https.


```
sub.example.com {
  proxy /syncthing 172.20.0.5:8384 172.21.0.3:8384
  redir /syncthing /syncthing/ 302
  transparent
  without /syncthing
}

[...]

2020/02/20 19:51:05 [INFO] Caddy version: v1.0.4
Activating privacy features... done.

Serving HTTP on port 2015
http://:2015

2020/02/20 19:51:05 [INFO] Serving http://:2015
2020/02/20 19:56:35 [ERROR] CaddyFile error: :128 - Error during parsing: Unknown directive 'transparent'
2020/02/20 19:56:35 [INFO] Wrong CaddyFile:

<caddyfile goes here>

```

**Comment by anLizard** [comment]
Commenting out the labels for syncthing made ports 80 and 443 work again

**Comment by lucaslorentz** [comment]
Oh sorry, I think it's my mistake...
It should be:
```
      - caddy.redir=/syncthing /syncthing/ 302
      - caddy.address=sub.example.com
      - caddy.sourcepath=/syncthing
      - caddy.targetport=8384
      - caddy.proxy.transparent=
      - caddy.proxy.without=/syncthing
```

**Comment by anLizard** [comment]
This configuration works! Thanks so much for this project, and for taking the time to help me out!
Now that I have this figured out as well I can fully migrate from nginx to caddy-proxy.

**Comment by anLizard** [comment]
Unfortunately, this method seems to break when trying to proxy multiple containers that need a similar redirect to the same subdomain, but on different paths, e.g. `sub.example.com/syncthing` and `sub.example.com/portainer`

I used the same labels on a second container and adjusted their values accordingly, but only one of the two containers is served as intended, while the other will always return 404.

If I remove the labels for Syncthing here, Portainer works. I can't get both to be proxied on the same subdomain with different paths

```
  syncthing:
    image: linuxserver/syncthing
    container_name: syncthing-0
    networks:
      - internal
      - public
    ports:
      #webUI
       #- 24100:8384
      #listening
      - 22000:22000
      #port discovery
      - 21027:21027/udp
    restart: always
    volumes:
      - ./syncthing/config:/config
      - ./syncthing/data:/data
      - /etc/localtime:/etc/localtime:ro
    environment:
      - UMASK_SET=<022>
      - PUID=1002
      - PGID=999
	labels:
      - caddy.redir=/syncthing /syncthing/ 302
      - caddy.address=sub.example.com
      - caddy.targetport=8384
      - caddy.sourcepath=/syncthing
      - caddy.proxy.transparent=
      - caddy.proxy.without=/syncthing
	  
  portainer:
    image: portainer/portainer
    container_name: portainer-0
    networks:
      - internal
    restart: always
    volumes:
      - ./portainer:/data
      - /etc/localtime:/etc/localtime:ro
      - /var/run/docker.sock:/var/run/docker.sock
    labels:
      - caddy.redir=/portainer /portainer/ 302
      - caddy.address=sub.example.com
      - caddy.targetport=9000
      - caddy.sourcepath=/portainer
      - caddy.proxy.transparent=
      - caddy.proxy.without=/portainer
```

The in-memory Caddyfile only creates the entry needed for Syncthing:

```
shh.lzrd.org {
  proxy /syncthing 172.20.0.5:8384 172.21.0.3:8384 {
    transparent
    without /syncthing
  }
  redir /syncthing /syncthing/ 302
}
```

I tried using `caddy_0:[...]` for Syncthing and `caddy_1:[...]` for Portainer, but got the same result

**Comment by lucaslorentz** [comment]
Can you please try again with version `lucaslorentz/caddy-docker-proxy:0.3.6`?
I improved it to keep both proxies you have because they have different source path.
No need to add _0 or _1.

Another alternative, that I usually do, is to put path in the address. Like:
```
caddy.address=sub.example.com/syncthing
```
That generates something like:
```
sub.example.com/syncthing {
  proxy / 172.20.0.5:8384
}
```
But that way I don't think you can do the redirect you need.

**Comment by lucaslorentz** [comment]
Seems that the redir is still not merged correctly.

**Comment by anLizard** [comment]
It is solved with 0.3.6!
Thanks so much once again!
